coordinator/src/main/scala/filodb.coordinator/ShardManager.scala

```diff
@@ -508,59 +518,66 @@ private[coordinator] final class ShardManager(settings: FilodbSettings,
+    logger.info(s"Assigning shards for dataset=$dataset to coordinator $coord for shards $shards")
     for { shard <- shards }  {
       val event = ShardAssignmentStarted(dataset, shard, coord)
       updateFromShardEvent(event)
     }
-    publishSnapshot(dataset)
-    /* If no shards are assigned to a coordinator, no commands are sent. */
-    logger.info(s"Sending start ingestion message for dataset=$dataset to coordinator $coord for shards $shards")
-    for {shard <- shards} coord ! StartShardIngestion(dataset, shard, None)
   }
 
-  /**
-    * This method has the shared logic for sending shard un-assignment messages
-    * to the coordinator, updating state for the event and broadcast of the state change to subscribers
-    */
-  private def sendUnassignmentMessagesAndEvents(dataset: DatasetRef,
-                                                coordinator: ActorRef,
-                                                mapper: ShardMapper,
-                                                shardsToDown: Seq[Int],
-                                                nodeUp: Boolean = true): Unit = {
-    logger.info(s"Sending stop ingestion message for dataset=$dataset to " +
+  private def doUnassignShards(dataset: DatasetRef,
+                               coordinator: ActorRef,
+                               shardsToDown: Seq[Int]): Unit = {
+    logger.info(s"Unassigning shards for dataset=$dataset to " +
       s"coordinator $coordinator for shards $shardsToDown")
     for { shard <- shardsToDown } {
       val event = ShardDown(dataset, shard, coordinator)
       updateFromShardEvent(event)
-      if (nodeUp) coordinator ! StopShardIngestion(dataset, shard)
     }
-    publishSnapshot(dataset)
-    logger.debug(s"Unassigned shards $shardsToDown of dataset=$dataset from $coordinator")
+  }
+
+  /**
+    * To be called after making a bunch of changes to the ShardMapper for the given dataset.
+    * Calling this method more often is permitted, but it generates more publish messages
+    * than is necessary.
+    */
+  private def publishChanges(ref: DatasetRef): Unit = {
+    publishSnapshot(ref)
+    updateShardMetrics()
   }
 
   /** Publishes a ShardMapper snapshot of given dataset to all subscribers of that dataset. */
-  def publishSnapshot(ref: DatasetRef): Unit =
+  def publishSnapshot(ref: DatasetRef): Unit = {
     mapperOpt(ref) match {
-      case Some(snapshot) =>
+      case Some(snapshot) => {
         for {
           subscription <- _subscriptions.subscription(ref)
         } subscription.subscribers foreach (_ ! snapshot)
+
+        // Also send a complete state command to all ingestion actors. Most are expected to
+        // actually be ingesting something, so also send the dataset info. This might be
+        // redundant, but sending it again is harmless.
+
+        val state = _datasetInfo(snapshot.ref)
+        val setupMsg = client.IngestionCommands.DatasetSetup(state.dataset.asCompactString,
```

Comment:
SetupMessage is supposed to be a one-time thing, but this method is going to be called from multiple places and definitely more than once in the lifetime of a coordinator for that dataset.   Creating the DatasetSetup message and serializing it is kinda expensive.   That is why a warning is logged when it is received and the dataset is already setup.  Let's chat and think about ways to not send this with every publishChanges()....

Reply:
How frequently are changes to be expected?

Reply:
I'm not sure having a large number of actors simultaneously reading from C* is a good idea.   I think having the `NodeClusterActor` send the setup message one time (when the datasets are setup) to all coordinators is probably better.

Some of this might need to be rethought also when we do more schema changes.